### PR TITLE
[wip] FWaaS v2.0 Rule Create

### PR DESCRIPTION
**Pull request by Elethiomel**
For #514

https://github.com/openstack/neutron-fwaas/blob/27906d0acff349c9e5ac7955d203fb2233c295db/neutron_fwaas/db/firewall/v2/firewall_db_v2.py#L68-L82

### CODE REVIEW

File: openstack/networking/v2/extensions/fwaas_v2/rules/results.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	Protocol             string `json:"protocol"`
+	Action               string `json:"action"`
+	IPVersion            int    `json:"ip_version,omitempty"`
+	SourceIPAddress      string `json:"source_ip_address,omitempty"`
+	DestinationIPAddress string `json:"destination_ip_address,omitempty"`
+	SourcePort           string `json:"source_port,omitempty"`
+	DestinationPort      string `json:"destination_port,omitempty"`
+	Shared               bool   `json:"shared,omitempty"`
+	Enabled              bool   `json:"enabled,omitempty"`
+	PolicyID             string `json:"firewall_policy_id"`
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/fwaas_v2/rules/results.go]: `FirewallPolicyID`

File: openstack/networking/v2/extensions/fwaas_v2/rules/results.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	Position             int    `json:"position"`
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/fwaas_v2/rules/results.go]: I'm unable to tell how this field is added to the result here: https://github.com/openstack/neutron-fwaas/blob/27906d0acff349c9e5ac7955d203fb2233c295db/neutron_fwaas/db/firewall/v2/firewall_db_v2.py#L269

Can you verify this field is really part of the JSON result?

File: openstack/networking/v2/extensions/fwaas_v2/rules/requests.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+)
+
+type CreateOptsBuilder interface {
+	ToRuleCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts contains all the values needed to create a new firewall rule.
+type CreateOpts struct {
+	Protocol             Protocol              `json:"protocol" required:"true"`
+	Action               string                `json:"action" required:"true"`
```
[the remark this answers is not in the document]
Comment (Elethiomel) [reply openstack/networking/v2/extensions/fwaas_v2/rules/requests.go]: Great idea :) I've added this in commit c677015065760efe6424966e3df2d3c71130f1c6

File: openstack/networking/v2/extensions/fwaas_v2/rules/results.go (hunk 1 above)
Comment (Elethiomel) [reply openstack/networking/v2/extensions/fwaas_v2/rules/results.go]: I've made this change in commit 19de2bdfcda05ef79ea686f5285188f0e488f1cd

File: openstack/networking/v2/extensions/fwaas_v2/rules/results.go (hunk 2 above)
Comment (Elethiomel) [reply openstack/networking/v2/extensions/fwaas_v2/rules/results.go]: Well spotted! I've verified that it's not. Removed in commit f1ad97effda2c92edd871a4728da1ce7c2f22479

Comment (Elethiomel) [reply openstack/networking/v2/extensions/fwaas_v2/rules/results.go]: The JSON result is also returning project_id, so I've added this in commit be00d44db04b8dd9fb72ed3acb2a66687f39f4c6

**Comment by Elethiomel** [comment]
@jtopjian commit ee2fb653f57c4d0bf1dba8ada013d1b10192e06f introduces an acceptance test for creating a rule, but the test is currently failing with a rather odd error... Any suggestion appreciated :)

`$ go test -tags fwaas_v2  .
--- FAIL: TestRuleC (0.60s)
    fwaas_v2.go:24: Attempting to create rule TESTACC-6t7dLduz with source 192.168.1.73:73 and destination 192.168.2.73:73
    fwaas_v2.go:42: Rule TESTACC-6t7dLduz successfully created
    convenience.go:35: Failure in fwaas_v2.go, line 45: expected "tcp" but got "tcp"
FAIL
FAIL    github.com/gophercloud/gophercloud/acceptance/openstack/networking/v2/extensions/fwaas_v2       0.601s
`

**Comment by jtopjian** [comment]
Nice work :)

The error is because you're comparing a string "tcp" with the Protocol type "tcp". There are two ways to resolve this:

```go
th.AssertEquals(t, rule.Protocol, "tcp")
th.AssertEquals(t, rule.Action, "allow")
```

or 

```go
th.AssertEquals(t, rule.Protocol, string(rules.ProtocolTCP))
th.AssertEquals(t, rule.Action, string(rules.ActionAllow))  
```

**Comment by Elethiomel** [comment]
@jtopjian Whoops! Thanks for the help :) Fixed in https://github.com/gophercloud/gophercloud/pull/1768/commits/f2ca260e5a49b1ba13e9b56fce226d1b3f721f2d

**Comment by Elethiomel** [comment]
@jtopjian What's the etiquette here in gophercloud dev land? Should I close off the changes requested or should I leave it to you?

**Comment by jtopjian** [comment]
> What's the etiquette here in gophercloud dev land? Should I close off the changes requested or should I leave it to you?

That's all on me. Sorry for the delay.

**Pull request merged**